src/main/java/com/artipie/npm/proxy/NpmProxy.java

```diff
@@ -83,130 +59,64 @@
-                                ).as(
-                                    BodyCodec.pipe(asyncfile)
-                                ).rxSend().flatMapMaybe(
-                                    response -> {
-                                        // @checkstyle MagicNumberCheck (1 line)
-                                        if (response.statusCode() == 200) {
-                                            return this.storage.save(
-                                                new NpmAsset(
-                                                    path,
-                                                    new RxFile(
-                                                        tmp.toPath(),
-                                                        this.vertx.fileSystem()
-                                                    ).flow(),
-                                                    response.getHeader("Last-Modified"),
-                                                    response.getHeader("Content-Type")
-                                                )
-                                            ).andThen(
-                                                Maybe.defer(
-                                                    () -> this.storage.getAsset(path)
-                                                )
-                                            );
-                                        } else {
-                                            Logger.debug(
-                                                this,
-                                                "Could not load asset: status code %d",
-                                                response.statusCode()
-                                            );
-                                            return Maybe.empty();
-                                        }
-                                    }
-                                ).onErrorResumeNext(
-                                    throwable -> {
-                                        Logger.error(
-                                            this,
-                                            "Error  occurred when process get asset call: %s",
-                                            throwable.getMessage()
-                                        );
-                                        return Maybe.empty();
-                                    }
-                                ).doOnTerminate(tmp::delete)
-                        );
-                    }
-                )
-            );
+        return this.storage.getAsset(path).switchIfEmpty(
+            Maybe.defer(
+                () -> this.vertx.fileSystem().rxCreateTempFile("npm-asset-", ".tmp")
+                    .flatMapMaybe(
+                        tmp -> this.remote.loadAsset(path, Paths.get(tmp)).flatMap(
+                            asset -> this.storage.save(asset)
```

Comment:
@Vatavuk no, it will fail. Asset has ByteBuffer publisher and it can not be re-winded after use to save it in the storage. We can re-read it from temp file (and duplicate all data fill logic) or re-read it from storage and I have suggested that the last one is better.